src/firejail/fs_bin.c

```diff
@@ -95,20 +95,34 @@ static char *check_dir_or_file(const char *name) {
 }
 
 static void duplicate(char *fname, FILE *fplist) {
-	if (*fname == '~' || *fname == '/' || strstr(fname, "..")) {
+	if (*fname == '~' || strstr(fname, "..")) {
 		fprintf(stderr, "Error: \"%s\" is an invalid filename\n", fname);
 		exit(1);
 	}
 	invalid_filename(fname);
 
-	char *path = check_dir_or_file(fname);
-	if (!path)
-		return;
-
-	// expand path, just in case this is a symbolic link
 	char *full_path;
-	if (asprintf(&full_path, "%s/%s", path, fname) == -1)
-		errExit("asprintf");
+	if (*fname == '/') {
+		// If the absolute filename is indicated, directly use it. This
+		// is required for the following three cases:
+		//  - if user's $PATH order is not the same as the above
+		//    paths[] variable order
+		//  - if for example /usr/bin/which is a symlink to /bin/which,
+		//    because in this case the result is a symlink pointing to
+		//    itself due to the file name being the same.
+		//  - if user wants to add a binary, which is not in the above
+		//    paths[] variable
+		if (asprintf(&full_path, "%s", fname) == -1)
+			errExit("asprintf");
```

Comment:
I'd change the error messages (here and below) to something useful, so user can understand what's wrong on debugging.

Reply:
Hello,
A grep command {grep -rsF 'errExit("asprintf")' * | wc -l} returned 244 occurrences of errExit("asprintf") in 52 files of the project source code.
So, it would be appropriate to do it for the 244 occurrences. However, people who had forked any of these 52 files will have to update their fork. Such a big change should be done/approved by official Firejail maintainers.